### ISSUE TITLE
Async vs. Sync event handling

### ISSUE THREAD

**Issue by SergeyTeplyakov** [nightroman/PowerShellTraps#7]
Noticed interesting behavior that I can't understand.

PowerShell supports two types of event handlers: sync event handlers and async event handlers.

First one is a plain-vanilla .NET event handling and second is PowerShell-based full-featured-generic event handling. 

Here small example:

``` powershell
Add-Type @"
public class Foo
{
  public event System.EventHandler CustomEvent;

  public void RaiseCustomEvent()
  {
      System.Console.WriteLine("Before raising CustomEvent");
      if (CustomEvent != null) {CustomEvent(this, System.EventArgs.Empty);}
      System.Console.WriteLine("After raising CustomEvent");
  }
}
"@

$foo = New-Object -TypeName Foo

# Register sync event
$foo.add_CustomEvent({
    Write-Host "Another Custom Handler"
})

# Register for async event
Register-ObjectEvent $foo -EventName CustomEvent -Action {
    Write-Host "Async CustomEventHandler"
}

# Raising the event
$foo.RaiseCustomEvent()
```

Running this code lead to the following output:

```
Before raising CustomEvent
Another Custom Handler
After raising CustomEvent
Async CustomEventHandler
```

So far, so good. But if to change order of subscriptions: async first and sync latter, then the output would be different:

``` powershell
# Register for async event
Register-ObjectEvent $foo -EventName CustomEvent -Action {
    Write-Host "Async CustomEventHandler"
}

# Register sync event
$foo.add_CustomEvent({
    Write-Host "Another Custom Handler"
})

# Raising the event
$foo.RaiseCustomEvent()
```

The output in this case would be:

```
Before raising CustomEvent
Async CustomEventHandler
Another Custom Handler
After raising CustomEvent
```

Basically, async event is not async any more! "Async" event handler was ran synchronously as well.

Am I missing something?

**Comment by nightroman** [comment]
I can reproduce this and I also find this strange. I do not have any explanation. Perhaps it's worth to ask at Stack Overflow as well. I will continue to investigate and then will add this to the traps. This fact is useful to know, regardless of an explanation.

Note that, unlike a PS sync handler, a C# sync handler added after the async event works as expected: sync works as sync, async as async.

**Comment by nightroman** [comment]
It is not that simple. It may have something to do with logging (`Write-Host` and `Console.WriteLine`). If I change the logging then I see different results (some expected and some new strange). To be continued.

**Comment by nightroman** [comment]
On other machines the issue is not stable. So it is either the race condition or
some wrong assumptions on how `Write-Host` and `Console` work in this async
scenario.

Here is a more carefully designed test. It does not show the problem, it works
as expected. If it works for you fine as well then the issue can be closed, no
action is needed. It was a useful exercise anyway.

``` powershell
Add-Type @'
using System;
using System.Collections;
public class TestEvent
{
    public bool Done;
    public event EventHandler CustomEvent;
    public void RaiseCustomEvent()
    {
        Log("Before raising CustomEvent");
        if (CustomEvent != null) {CustomEvent(this, EventArgs.Empty);}
        Log("After raising CustomEvent");
    }
    public void Log(string text)
    {
        lock(_lock)
        {
            _log.Add(text);
        }
    }
    public object[] Log()
    {
        return _log.ToArray();
    }
    object _lock = new object();
    ArrayList _log = new ArrayList();
}
'@

$object = New-Object -TypeName TestEvent

# Register for async event
$null = Register-ObjectEvent $object -EventName CustomEvent -Action {
    Start-Sleep 2 # ensure it does not work before sync
    $Sender.Log('Async handler')
    $Sender.Done = $true
}

# Register sync event
$object.add_CustomEvent({
    $this.Log('Sync handler')
})

# Raising the event
$object.RaiseCustomEvent()

while(!$object.Done) {}
$object.Log()
```

**Comment by nightroman** [comment]
P.S. Even the above more careful test may show "unexpected" order of log records. It happened when I tested it the first time in PS v2. Perhaps due to the cold start of it 2 seconds in `Start-Sleep 2` was not enough to ensure the expected order. Next runs in v2 show the expected" order. But technically this is not guaranteed.

**Comment by SergeyTeplyakov** [comment]
Sorry, didn't close this issue myself. Thanks a lot for explanation! now it makes more sense!
